lib/Selenium/Remote/Driver/ActionChains.pm

```diff
@@ -0,0 +1,155 @@
+
+has 'driver' => ( 
+    is => 'rw', 
+);
+
+has 'actions' => ( 
+    is => 'lazy', 
+    builder => sub { [] },
+    clearer => 1,
+);
+
+sub perform { 
+    my $self = shift; 
+    foreach my $action (@{$self->actions}) { 
+        $action->();
+    }
+}
+
+sub click { 
+    my $self = shift; 
+    my $element = shift; 
+    if ($element) { 
+       $self->move_to_element($element); 
+    }
+    # left click
+    push @{$self->actions}, sub { $self->driver->click('LEFT') };
+    $self; 
+}
+
+sub click_and_hold { 
+    my $self = shift; 
+    my $element = shift; 
+    if ($element) { 
+       $self->move_to_element($element); 
+    }
+    push @{$self->actions}, sub { $self->driver->button_down };
+    $self; 
+}
+
+sub context_click { 
+    my $self = shift; 
+    my $element = shift; 
+    if ($element) { 
+       $self->move_to_element($element); 
+    }
+    # right click
+    push @{$self->actions}, sub { $self->driver->click('RIGHT') }; 
+    $self; 
+}
+
+
+sub double_click { 
+    my $self = shift; 
+    my $element = shift; 
+    if ($element) { 
+       $self->move_to_element($element); 
+    }
+    push @{$self->actions}, sub { $self->driver->double_click };
+    $self; 
+}
+
+sub release { 
+    my $self = shift; 
+    my $element = shift; 
+    if ($element) { 
+       $self->move_to_element($element); 
+    }
+    push @{$self->actions}, sub { $self->driver->button_up };
+    $self; 
+}
+
+sub drag_and_drop { 
+    my $self = shift; 
+    my ($source,$target) = @_;
+    $self->click_and_hold($source);
+    $self->release($target);
+    $self; 
+}
+
+sub drag_and_drop_by_offset { 
+    my $self = shift; 
+    my ($source,$xoffset,$yoffset) = @_;
+    $self->click_and_hold($source); 
+    $self->move_by_offset($xoffset,$yoffset);
+    $self->release($source); 
+    $self;
+}
+
+sub move_to_element {
+    my $self    = shift;
+    my $element = shift;
+    push @{ $self->actions },
+      sub { $self->driver->move_to( element => $element ) };
+    $self;
+}
+
+sub move_by_offset {
+    my $self = shift;
+    my ( $xoffset, $yoffset ) = @_;
+    push @{ $self->actions }, sub {
+        $self->driver->move_to( xoffset => $xoffset, yoffset => $yoffset );
+    };
+    $self;
+}
+
+sub move_to_element_with_offset {
+    my $self = shift;
+    my ( $element, $xoffset, $yoffset ) = @_;
+    push @{ $self->actions }, sub {
+        $self->driver->move_to( element => $element, xoffset => $xoffset,
+            yoffset => $yoffset );
+    };
+    $self;
+}
+
+sub key_down { 
+    my $self = shift; 
+    my ($value ,$element) = @_;
+    if (defined($element)) { 
+        $self->click($element);
+    }
+    push @{ $self->actions }, sub { $self->driver->send_keys_to_active_element(@$value) };
+    $self;
+}
+
+sub key_up { 
+    my $self = shift; 
+    my ($value ,$element) = @_;
+    if (defined($element)) { 
+        $self->click($element);
+    }
+    push @{ $self->actions }, sub { $self->driver->send_keys_to_active_element(@$value) };
```

Comment:
I had to get my mind around it too. 
First, I sticked to the [Python implementation](http://selenium.googlecode.com/svn/trunk/docs/api/py/_modules/selenium/webdriver/common/action_chains.html#ActionChains), which is roughly the same, but I think  we can make one generic function of those two. 
Then, there is the fact that calling `send_keys_to_active_element` with no prior call is like pressing down a button without releasing (when you want to send a modifier, like Alt, Shift, Ctrl), so you have to call `send_keys_to_active_element` with the same key a second time to release the key. 
Because of this, it kind of make sense to have two functions to make the difference between the twos.